### PR TITLE
Include directories when using a S3Client to list objects in a prefix.

### PR DESCRIPTION
**Pull request by simtel12**
Any of the checked boxes below indicate that I took action:

- [X] Reviewed the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md#contribution-workflow).
- [X] Signed the [Gradle CLA](http://gradle.org/contributor-license-agreement/).
- [X] Ensured that basic checks pass: `./gradlew quickCheck`

For all non-trivial changes that modify the behavior or public API:

- [ ] Before submitting a pull request, I started a discussion on the
[Gradle developer list](https://groups.google.com/forum/#!forum/gradle-dev)
or can reference a [JIRA issue](https://issues.gradle.org/secure/Dashboard.jspa).
- [ ] I considered writing a design document. A design document can be
brief but explains the use case or problem you are trying to solve,
touches on the planned implementation approach as well as the test cases
that verify the behavior. Optimally, design documents should be submitted
as a separate pull request. [Samples](https://github.com/gradle/gradle/tree/master/design-docs)
can be found in the Gradle GitHub repository. Please let us know if you need help with
creating the design document. We are happy to help!
- [X] The pull request contains an appropriate level of unit and integration
test coverage to verify the behavior. Before submitting the pull request
I ran a build on your local machine via the command
`./gradlew quickCheck <impacted-subproject>:check`.
- [ ] The pull request updates the Gradle documentation like user guide,
DSL reference and Javadocs where applicable.

Ensure that directories are correctly included when listing the contents of an S3 prefix so that dynamic version dependencies will be able to discover the correct version to use.

- Directories are given as 'common prefixes' when listing the the items in an S3 prefix, so we need to process them into the list of things returned by the S3Client.

### CODE REVIEW

**Comment by bmuschko** [comment]
Thanks for providing the pull request. Could you please give me the name or email you used for signing the CLA? I couldn't find one associated with your name.

**Comment by simtel12** [comment]
To confirm: were you able to find my CLA?  I notice that the comment asking for it is gone.

**Comment by bmuschko** [comment]
Yes, I found it. Thanks for checking in with us.

**Comment by wolfs** [comment]
Thank you for your contribution!

It is not clear to me what problem this pull request is trying to solve. Could you create an issue or start a discussion on the mailing list?

File: subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java [hunk 1]
```diff
@@ -42,6 +42,7 @@
 public class S3Client {
     private static final Logger LOGGER = LoggerFactory.getLogger(S3Client.class);
     private static final Pattern FILENAME_PATTERN = Pattern.compile("[^\\/]+\\.*$");
+    private static final Pattern DIRNAME_PATTERN = Pattern.compile("([^\\/]+)[\\/]$");
```
Comment (wolfs) [inline subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java]: Please remove unnecessary escaping and character groups from the regular expressions.

File: subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java [hunk 2]
```diff
@@ -153,10 +154,31 @@ public S3Object getResource(URI uri) {
                     results.add(fileName);
                 }
             }
+
+        }
+
+        List<String> commonPrefixes = objectListing.getCommonPrefixes();
```
Comment (wolfs) [inline subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java]: Please extract a method for getting all the directory names from the objectListing as well as for the file names.

**Comment by wolfs** [comment]
Please address my comments in order to get this merged.

File: subprojects/resources-s3/src/test/groovy/org/gradle/internal/resource/transport/aws/s3/S3ClientTest.groovy [hunk 3]
```diff
@@ -98,8 +98,17 @@ class S3ClientTest extends Specification {
         objectListing.getCommonPrefixes() >> ['/SNAPSHOT/', '/SNAPSHOT/1.0.8/']
 
         when:
-        def results = s3Client.resolveResourceNames(objectListing)
+        def results = s3Client.resolveDirectoryResourceNames(objectListing)
```
Comment (wolfs) [inline subprojects/resources-s3/src/test/groovy/org/gradle/internal/resource/transport/aws/s3/S3ClientTest.groovy]: The test exercises private methods. Please move the methods resolveResourceNames, resolveDirectoryResourceNames and resolveFileResourceNames to a new object (e.g. S3ResourceResolver), have them public there and then have the Test test the new class.

File: subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java [hunk 4]
```diff
@@ -187,7 +187,7 @@ private String extractDirectoryName(String key) {
         Matcher matcher = DIRNAME_PATTERN.matcher(key);
         if (matcher.find()) {
             // Add a "/" to differentiate directories from files
-            return matcher.group(1) + "/";
+            return matcher.group(0);
```
Comment (wolfs) [inline subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java]: The comment does not make much sense now, does it? Please change it or remove it.

File: subprojects/resources-s3/src/test/groovy/org/gradle/internal/resource/transport/aws/s3/S3ClientTest.groovy [hunk 5]
```diff
@@ -80,11 +95,22 @@ class S3ClientTest extends Specification {
         objectSummary2.getKey() >> '/SNAPSHOT/someOther.jar'
         objectListing.getObjectSummaries() >> [objectSummary, objectSummary2]
 
+        objectListing.getCommonPrefixes() >> ['/SNAPSHOT/', '/SNAPSHOT/1.0.8/']
+
         when:
-        def results = s3Client.resolveResourceNames(objectListing)
+        def results = s3Client.resolveDirectoryResourceNames(objectListing)
```
Comment (wolfs) [inline subprojects/resources-s3/src/test/groovy/org/gradle/internal/resource/transport/aws/s3/S3ClientTest.groovy]: The test exercises private methods. Please move the methods resolveResourceNames, resolveDirectoryResourceNames and resolveFileResourceNames to a new object (e.g. S3ResourceResolver), have them public there and then have the Test test the new class.

File: subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java [hunk 6]
```diff
@@ -154,9 +164,34 @@ public S3Object getResource(URI uri) {
...
+            }
+        }
+
         return results;
     }
 
+    private String extractDirectoryName(String key) {
+        Matcher matcher = DIRNAME_PATTERN.matcher(key);
+        if (matcher.find()) {
+            // Add a "/" to differentiate directories from files
```
Comment (wolfs) [inline subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java]: The comment does not make much sense now, does it? Please rephrase or remove it.

**Comment by wolfs** [comment]
I added some more comments. Please address them, too. Thank you for your patience.

**Comment by wolfs** [comment]
I created an issue for this bug: https://issues.gradle.org/browse/GRADLE-3453

**Comment by wolfs** [comment]
Please make the methods on S3ResourceResolver and S3ResourceNameExtractor non-static. Then have S3Client and S3ResourceResolver have an instance of the corresponding class and call methods on this one.

**Comment by simtel12** [comment]
I have no problem doing this, but I'm curious about the reasoning.  It seems like a needless allocation.


________________________________

**Comment by wolfs** [comment]
@simtel12: If the methods are not static then testing is easier. E.g. one could replace the S3ResourceResolver in the S3Client by a Mock and then only test the S3Client. The overhead for this is minimal, since there will not be many S3Clients.
Moreover, it is then difficult to change the behavior of S3ResourceResolver by creating a subclass, decorator, etc.
This is all not very probable for the methods at hand, but as a rule of thumb I would make the methods non-static and only make them static if there is a very good reason for it.

**Comment by simtel12** [comment]
@wolfs Do you think there's any chance that this might make it into 2.14?  I'm guessing not, since rc1 is already out, but it would be pretty awesome if it could.

**Comment by wolfs** [comment]
@simtel12: I am sorry, but this change will not make it into 2.14 since it is already too late.

**Comment by wolfs** [comment]
Internal review: https://code-review.gradle.org/cru/REVIEW-5988

**Comment by big-guy** [comment]
LGTM

**Pull request merged**